#### **Assignment 5 🛒🛍️** 

#### **Student Name:** Nadeen moreb 

#### **Student ID:** 1203437 

# Online Retail II UCI

### importing numpy and pandas libraries 

In [1]:
# importing numpy and pandas libraries 
import numpy as np 
import pandas as pd

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/online-retail-ii-uci/online_retail_II.csv


## Loading The Data

In [2]:
df = pd.read_csv("/kaggle/input/online-retail-ii-uci/online_retail_II.csv")

In [3]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3  2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4  2009-12-01 07:45:00   1.25      13085.0  United Kingdom

## Randomly select 200k row from the dataset using your ID as seed.

In [4]:
# assign seed to my id number = 1203437
seed = 1203437  

# set the random seed
np.random.seed(seed)

# Randomly select 200,000 rows from the DataFrame
random_sample = df.sample(n=200000)


### 1. Determine the total sales per invoice and identify the top 5 invoices.

In [5]:
# Compute total sales per invoice
random_sample["TotalSales"] = random_sample["Quantity"] * random_sample["Price"]

#identify the top 5 invoices
random_sample.groupby("Invoice").agg({"TotalSales": "sum"}).nlargest(5, columns="TotalSales")

TotalSales
Invoice            
531516     17660.16
524181     11674.64
502263     10953.50
496115      8985.60
493819      8432.68

#### these are the top 5 invoices in terms of total sales amount:

#### - Invoice 531516: has a total sales amount of £17,660.16.

#### - Invoice 524181: has a total sales amount of £11,674.64.

#### - Invoice 502263: has a total sales amount of £10,953.50.

#### - Invoice 496115: has a total sales amount of £8,985.60.

#### - Invoice 493819: has a total sales amount of £8,432.68.

### 2. Compute the total sales per customer and list the top 5 customers.

In [6]:
# Compute total sales per customer
random_sample["TotalSales"] = random_sample['Quantity'] * random_sample['Price']

#identify the top 5 customers
random_sample.groupby("Customer ID").agg({"TotalSales": "sum"}).nlargest(5, columns="TotalSales")

TotalSales
Customer ID            
18102.0       113616.94
14646.0        99891.48
17450.0        56917.96
14911.0        53998.25
14156.0        51950.51

#### these are the top 5 customers in terms of total sales amount:

#### - Customer 18102.0: has a total sales amount of £113616.94

#### - Customer 14646.0: has a total sales amount of £99891.48

#### - Customer 17450.0: has a total sales amount of £56917.96

#### - Customer 14911.0: has a total sales amount of £53998.25

#### - Customer 14156.0: has a total sales amount of £51950.51

### 3. Identify the top 5 selling items during this period.

In [7]:
# Compute total sales per customer
random_sample["TotalSales"] = random_sample['Quantity'] * random_sample['Price']

#identify the top 5 customers
random_sample.groupby("StockCode").agg({"TotalSales": "sum"}).nlargest(5, columns="TotalSales")





TotalSales
StockCode            
22423        60543.08
DOT          45752.15
85123A       40497.31
85099B       28530.55
47566        26223.50

#### - StockCode 22423: has total sales amount of £60,543.08.
#### - StockCode DOT: has total sales amount of £45,752.15.
#### - StockCode 85123A: has total sales amount of £40,497.31.
#### - StockCode 85099B: has total sales amount of £28,530.55.
#### - StockCode 47566: has total sales amount of £26,223.50.

### 4. Determine the most frequently purchased item pairs by customers and display the top 20 pairs.

In [8]:
from itertools import combinations

# لroup by Invoice and create a list of StockCodes for each invoice
invoice_items = random_sample.groupby('Invoice')['StockCode'].apply(list)

# Function that generates pairs of items from the list of items/produxts
def generate_pairs(item_list):
    return list(combinations(item_list, 2))

# call the function to generate pairs for each invoice, and then stack the pairs
pair_counts = invoice_items.apply(generate_pairs).explode().value_counts()

# print top 20 frequently purchased item pairs
top_pairs = pair_counts.head(20)
print("Top 20 frequently purchased item pairs are:")
print(top_pairs)


Top 20 frequently purchased item pairs are:
StockCode
(21733, 85123A)     33
(85099B, 22386)     33
(20725, 20727)      33
(20728, 22383)      30
(85099F, 85099B)    29
(85099B, 21931)     29
(22697, 22698)      28
(84991, 21212)      28
(20723, 20724)      28
(20725, 22383)      27
(20728, 20725)      26
(20725, 22384)      26
(21080, 21212)      26
(22910, 22086)      26
(22697, 22699)      26
(22383, 22384)      26
(21931, 85099B)     26
(21212, 21213)      25
(20724, 22356)      25
(22384, 20728)      25
Name: count, dtype: int64


the results show the top 20 frequently purchased item pairs with the count of how many times each pair appears in the dataset. 

- (21733, 85123A): This pair consists of items with StockCodes 21733 and 85123A, and it appears 33 times in the dataset.

- (85099B, 22386): This pair consists of items with StockCodes 85099B and 22386, and it appears 33 times in the dataset.

- (20725, 20727): This pair consists of items with StockCodes 20725 and 20727, and it appears 33 times in the dataset.
and so on...

### 5. The retail store aims to introduce a new promotion targeting items with low sales volume. This promotion entails offering a discounted price for a pair of two different items, with one item being among the top 10 best sellers and the other from the bottom 10 sellers. Customers who purchase these two items together will qualify for the special pricing. Your task is to recommend 10 pairs of items to kickstart this campaign.


In [11]:
# the top 10 least preferred products list
most_preferred_products = random_sample['Description'].value_counts()[:10]

# the bottom 10 least preferred products list
least_preferred_products = random_sample['Description'].value_counts().nsmallest(10)

# list of pairs
pair_list = [(most_preferred_products.index[i], least_preferred_products.index[i]) for i in range(min(len(most_preferred_products), len(least_preferred_products)))]

# print each pair
for pair in pair_list:
    print(pair[0], " , ", pair[1])


WHITE HANGING HEART T-LIGHT HOLDER  ,  SET OF 36 DOILIES VINTAGE CHRISTMAS
REGENCY CAKESTAND 3 TIER  ,  FLOWER FAIRY MUG+COASTER,ROSE
JUMBO BAG RED RETROSPOT  ,  DIAMANTE BOW BROOCH WHITE COLOUR
ASSORTED COLOUR BIRD ORNAMENT  ,  WRAP WINTER FOREST
PARTY BUNTING  ,  FLOWERS HANDBAG blue and orange
LUNCH BAG  BLACK SKULL.  ,  BLACK CRYSTAL DROP EARRINGS
STRAWBERRY CERAMIC TRINKET BOX  ,  SEAGRASS S/2 CORNER LAUNDRY 
JUMBO STORAGE BAG SUKI  ,  wrongly marked
REX CASH+CARRY JUMBO SHOPPER  ,  Wrongly mrked had 85123a in box
HEART OF WICKER SMALL  ,  WHITE VINT ART DECO CRYSTAL NECKLAC


these are pairs that consists of an item from 10 best sellers and an item from 10 bottom sellers
- WHITE HANGING HEART T-LIGHT HOLDER  ,  SET OF 36 DOILIES VINTAGE CHRISTMAS

- REGENCY CAKESTAND 3 TIER  ,  FLOWER FAIRY MUG+COASTER,ROSE

- JUMBO BAG RED RETROSPOT  ,  DIAMANTE BOW BROOCH WHITE COLOUR

- ASSORTED COLOUR BIRD ORNAMENT  ,  WRAP WINTER FOREST

and so on...